### PR TITLE
feat(files): move multipart functionality to stable [BND3D-4545]

### PR DESCRIPTION
**Pull request by G0nza1es**
We are promoting multi-part upload functionality to stable after using it for months in 3d-management. This will allow other consumers to use it without adding `beta-sdk `dependency in their apps.

### CODE REVIEW

**Comment by samir-ajdarpasic** [comment]
Shouldn't the feature be present in stable and beta ?

**Comment by G0nza1es** [comment]
> Shouldn't the feature be present in stable and beta ?

ideally yes. i have struggled to make it so. like Beta extends Stable 

I get type conflicts  when i make it that way in this current structure. is there a PR that i can follow as a guide? i know it is a breaking change for beta but we are the only consumers of it . Not sure what is the guideline how we retire beta functionality when we promote it to stable . i can have a quick chat if you have time on this

**Comment by samir-ajdarpasic** [comment]
> > Shouldn't the feature be present in stable and beta ?
> 
> ideally yes. i have struggled to make it so. like Beta extends Stable
> 
> I get type conflicts when i make it that way in this current structure. is there a PR that i can follow as a guide? i know it is a breaking change for beta but we are the only consumers of it . Not sure what is the guideline how we retire beta functionality when we promote it to stable . i can have a quick chat if you have time on this

@G0nza1es I'm not very familiar with this repo, haven't contributed here much, sorry :/
maybe @polomani or @BugGambit can help....

**Comment by BugGambit** [comment]
> > > Shouldn't the feature be present in stable and beta ?
> > 
> > 
> > ideally yes. i have struggled to make it so. like Beta extends Stable
> > I get type conflicts when i make it that way in this current structure. is there a PR that i can follow as a guide? i know it is a breaking change for beta but we are the only consumers of it . Not sure what is the guideline how we retire beta functionality when we promote it to stable . i can have a quick chat if you have time on this
> 
> @G0nza1es I'm not very familiar with this repo, haven't contributed here much, sorry :/ maybe @polomani or @BugGambit can help....

@G0nza1es the beta SDK extends the stable SDK. So as long as you don't break the implementation but just copy the API methods from beta to stable, then the beta SDK shouldn't break.

**Comment by G0nza1es** [comment]
> > > > Shouldn't the feature be present in stable and beta ?
> > > 
> > > 
> > > ideally yes. i have struggled to make it so. like Beta extends Stable
> > > I get type conflicts when i make it that way in this current structure. is there a PR that i can follow as a guide? i know it is a breaking change for beta but we are the only consumers of it . Not sure what is the guideline how we retire beta functionality when we promote it to stable . i can have a quick chat if you have time on this
> > 
> > 
> > @G0nza1es I'm not very familiar with this repo, haven't contributed here much, sorry :/ maybe @polomani or @BugGambit can help....
> 
> @G0nza1es the beta SDK extends the stable SDK. So as long as you don't break the implementation but just copy the API methods from beta to stable, then the beta SDK shouldn't break.

ok. initially i tried that but when i have the same types , i was having conflicting types. but will try one more time . maybe i was doing something wrong.

**Comment by G0nza1es** [comment]
> > > > > Shouldn't the feature be present in stable and beta ?
> > > > 
> > > > 
> > > > ideally yes. i have struggled to make it so. like Beta extends Stable
> > > > I get type conflicts when i make it that way in this current structure. is there a PR that i can follow as a guide? i know it is a breaking change for beta but we are the only consumers of it . Not sure what is the guideline how we retire beta functionality when we promote it to stable . i can have a quick chat if you have time on this
> > > 
> > > 
> > > @G0nza1es I'm not very familiar with this repo, haven't contributed here much, sorry :/ maybe @polomani or @BugGambit can help....
> > 
> > 
> > @G0nza1es the beta SDK extends the stable SDK. So as long as you don't break the implementation but just copy the API methods from beta to stable, then the beta SDK shouldn't break.
> 
> ok. initially i tried that but when i have the same types , i was having conflicting types. but will try one more time . maybe i was doing something wrong.

@BugGambit now done. not sure what went wrong last time but  working now. would be good to get feedback.

**Comment by G0nza1es** [comment]
@BugGambit @polomani would it possible to get some feedback on this ?

**Comment by polomani** [comment]
@G0nza1es I see you left the beta implementation. You should remove it, since there is an implementation in stable package now and beta extends stable anyway.

So, remove entire `packages/beta/src/api/files` folder

**Comment by G0nza1es** [comment]
> @G0nza1es I see you left the beta implementation. You should remove it, since there is an implementation in stable package now and beta extends stable anyway.
> 
> So, remove entire `packages/beta/src/api/files` folder

@polomani  done. I yalc -ed both beta and stable in my local machine and  no type errors 👍🏽

File: packages/stable/src/api/files/filesApi.ts [hunk 1]
```diff
@@ -233,6 +248,52 @@ export class FilesAPI extends BaseResourceAPI<FileInfo> {
     );
     return this.addToMapAndReturn(response.data.items, response);
   }
+  /**
+   * [Init a multipart file upload](https://api-docs.cognite.com/20230101/tag/Files/operation/initMultiPartUpload)
+   *
+   * ```js
+   * const numberOfParts = 5;
+   * // automatic upload:
+   * const multiPartUploadApi = await client.files.upload({name: 'examplefile.jpg', mimeType: 'image/jpeg'}, numberOfParts);
```
Comment (polomani) [inline packages/stable/src/api/files/filesApi.ts]: is this doc string a valid code?

Comment (polomani) [reply packages/stable/src/api/files/filesApi.ts]: I guess it should be
```
const multiPartUpload = await client.files.multipartUploadSession(...)
```

Comment (G0nza1es) [reply packages/stable/src/api/files/filesApi.ts]: yup. good catch. fixed

**Review by polomani — DISMISSED** [review]
good on my side, but only did a brief review (and I am not a code owner)

**Comment by BugGambit** [comment]
@G0nza1es FYI, this is a breaking change to the beta SDK. I am working on releasing v10, which will include a new major version for all packages.
Can I suggest you change the base branch for this PR from `master` to `release-v10`?

File: packages/stable/src/api/files/filesApi.ts [hunk 2]
```diff
@@ -21,9 +23,22 @@ import type {
...
+    minimumNumberOfParts: 1,
+    maxNumberOfParts: 250,
+  };
+  private _client: CDFHttpClient;
+  private _map: MetadataMap;
+  private _baseUrl: string;
+  constructor(...args: [string, CDFHttpClient, MetadataMap]) {
+    super(...args);
+    [this._baseUrl, this._client, this._map] = args;
+  }
```
Comment (BugGambit) [inline packages/stable/src/api/files/filesApi.ts:41]:
```suggestion
  constructor(private _client: CDFHttpClient, private _map: MetadataMap, private _baseUrl: string) {
    super(...args);
  }
```

File: packages/stable/src/api/files/filesApi.ts [hunk 3]
```diff
@@ -233,6 +248,52 @@ export class FilesAPI extends BaseResourceAPI<FileInfo> {
...
+  private async getMultipartUploadSession(
+    fileInfo: ExternalFileInfo,
+    parts: number,
+    overwrite = false
+  ) {
+    if (
+      parts < this.limits.minimumNumberOfParts ||
+      parts > this.limits.maxNumberOfParts
+    ) {
+      throw Error('parts must be greater than 0 and less than 250');
```
Comment (BugGambit) [inline packages/stable/src/api/files/filesApi.ts]: Suggestion: don't hard code dynamic values in the string. If we update limits, then the error message should reflect that.
```suggestion
      throw Error(`parts must be in range ${this.limits.minimumNumberOfParts} <= parts <= and less than ${this.limits.maxNumberOfParts}`);
```

File: packages/stable/src/__tests__/testUtils.ts [hunk 4]
```diff
@@ -186,6 +183,99 @@ const getFileCreateArgs = (
...
+      login().then((account) => {
+        return account.access_token;
+      }),
+  });
+}
+
+export function setupMockableClientForUnitTest() {
+  const client = setupLoggedInClientForUnitTest();
+  return client;
+}
```
Comment (BugGambit) [inline packages/stable/src/__tests__/testUtils.ts]: This function doesn't nothing? It's just a proxy to `setupLoggedInClientForUnitTest`

File: packages/stable/src/__tests__/api/filesMultiPart.int.spec.ts [hunk 5]
```diff
@@ -1,5 +1,3 @@
-// Copyright 2020 Cognite AS
```
Comment (msusag) [inline packages/stable/src/__tests__/api/filesMultiPart.int.spec.ts]: I see the copyright is kept in some files, and removed in some.
Did you intend to delete the copyright comment?

File: packages/stable/src/api/files/filesApi.ts [hunk 6]
```diff
@@ -233,6 +248,52 @@ export class FilesAPI extends BaseResourceAPI<FileInfo> {
     );
     return this.addToMapAndReturn(response.data.items, response);
   }
+  /**
+   * [Init a multipart file upload](https://api-docs.cognite.com/20230101/tag/Files/operation/initMultiPartUpload)
+   *
+   * ```js
+   * const numberOfParts = 5;
```
Comment (msusag) [inline packages/stable/src/api/files/filesApi.ts:256]: Probably worth noting in the example here that the number of parts needs to be carefully calculated, to adhere to the part size requirements in the API documentation, which states:

> Each file part must be larger than 5 MiB, and smaller than 4000 MiB. The file part for the last uploadURL can be smaller than 5 MiB.

File: packages/stable/src/__tests__/testUtils.ts (hunk 4 above)
Comment (G0nza1es) [reply packages/stable/src/__tests__/testUtils.ts]: yup deleted

File: packages/stable/src/__tests__/api/filesMultiPart.int.spec.ts (hunk 5 above)
Comment (G0nza1es) [reply packages/stable/src/__tests__/api/filesMultiPart.int.spec.ts]: not sure but added back

File: packages/stable/src/api/files/filesApi.ts (hunk 6 above)
Comment (G0nza1es) [reply packages/stable/src/api/files/filesApi.ts:256]: added that comment

File: packages/stable/src/api/files/filesApi.ts (hunk 2 above)
Comment (G0nza1es) [reply packages/stable/src/api/files/filesApi.ts:41]: unfortunately this doesnt work :(

**Comment by G0nza1es** [comment]
> @G0nza1es FYI, this is a breaking change to the beta SDK. I am working on releasing v10, which will include a new major version for all packages. Can I suggest you change the base branch for this PR from `master` to `release-v10`?

I can do that but as beta extends stable.  when this is merged new beta users should be able to use sdk without any breaking change. `beta-sdk `has a `stable-sdk` dependency . am i wrong to think like that ? if so i can rebase it to that branch

File: packages/stable/src/__tests__/api/datamodels.int.spec.ts [hunk 7]
```diff
@@ -18,7 +18,7 @@ describe('Data models integration test', () => {
   const datamodelCreationDefinition: DataModelCreate = {
     externalId: `test_datamodel_${randomInt()}`,
     space: TEST_SPACE_NAME,
-    name: 'test_datamodel',
```
Comment (msusag) [inline packages/stable/src/__tests__/api/datamodels.int.spec.ts]: Any reason for why this name had to be changed?

File: packages/stable/src/__tests__/api/filesMultiPart.unit.spec.ts [hunk 8]
```diff
@@ -64,7 +66,9 @@ describe('Multi part upload unit test', () => {
           numberOfParts,
           true
         )
-      ).rejects.toThrowError('parts must be greater than 0 and less than 250');
+      ).rejects.toThrowError(
+        'parts must be in range 1 <= parts <= and less than 250'
```
Comment (msusag) [inline packages/stable/src/__tests__/api/filesMultiPart.unit.spec.ts]:
```suggestion
        'parts must be in range 1 <= parts <= 250'
```

File: packages/stable/src/__tests__/testUtils.ts [hunk 9]
```diff
@@ -186,6 +183,94 @@ const getFileCreateArgs = (
 
   return { postfix, fileContent, sourceCreatedTime, localFileMeta };
 };
+export function setupMockableClientForIntegrationTests() {
+  const client = setupClient(process.env.COGNITE_BASE_URL);
+  return client;
+}
+export function setupLoggedInClientForUnitTest() {
+  return new CogniteClient({
+    appId: 'JS SDK integration tests (beta)',
```
Comment (msusag) [inline packages/stable/src/__tests__/testUtils.ts]: The appId should perhaps not include the "(beta)" keyword in the name now, as it is moved to stable?

File: packages/stable/src/__tests__/testUtils.ts [hunk 10]
```diff
@@ -186,6 +183,94 @@ const getFileCreateArgs = (
...
+  // Divide the file content into chunks
+  for (let i = 0; i < numChunks; i++) {
+    const start = i * chunkSize;
+    const end = start + chunkSize;
+    chunks.push(fileContentBinary.slice(start, end));
+  }
+
+  return chunks;
+}
+function toArrayBuffer(buffer: Buffer) {
```
Comment (msusag) [inline packages/stable/src/__tests__/testUtils.ts:225]:
```suggestion

function toArrayBuffer(buffer: Buffer) {
```

Comment (msusag) [reply packages/stable/src/__tests__/testUtils.ts:225]: Just add a empty newline between functions.

Comment (msusag) [reply packages/stable/src/__tests__/testUtils.ts:225]: Or is empty newline between functions not the norm in this file?

File: packages/stable/src/api/files/filesApi.ts [hunk 11]
```diff
@@ -233,6 +248,55 @@ export class FilesAPI extends BaseResourceAPI<FileInfo> {
...
+    fileInfo: ExternalFileInfo,
+    parts: number,
+    overwrite = false
+  ) {
+    if (
+      parts < this.limits.minimumNumberOfParts ||
+      parts > this.limits.maxNumberOfParts
+    ) {
+      throw Error(
+        `parts must be in range ${this.limits.minimumNumberOfParts} <= parts <= and less than ${this.limits.maxNumberOfParts}`
```
Comment (msusag) [inline packages/stable/src/api/files/filesApi.ts]:
```suggestion
        `parts must be in range ${this.limits.minimumNumberOfParts} <= parts <= ${this.limits.maxNumberOfParts}`
```

File: packages/stable/src/api/files/filesMultipartUploadSessionApi.ts [hunk 12]
```diff
@@ -22,7 +22,7 @@ export class FilesMultipartUploadSessionAPI extends BaseResourceAPI<FileInfo> {
     [, , , this.multiPartFileUploadResponse] = args;
 
     this.uploadedUrls = this.multiPartFileUploadResponse.uploadUrls.map(
-      () => false
+      (_) => false
     );
     this.finished = false;
   }
```
Comment (msusag) [inline packages/stable/src/api/files/filesMultipartUploadSessionApi.ts:28]: Error text on line 36 should be:
   throw Error('you are uploading empty content');

Error text on line 42 should be:
    throw Error('part number is outside allowed range 0 to 249');

Ideally the numbers there should in stead of being hardcoded be taken from the constants defined in FilesApi.limits, using the limit - 1.

**Review by msusag — APPROVED** [review]
Looks good to me. 
Added some minor comments and suggestions.

**Comment by G0nza1es** [comment]
> > @G0nza1es FYI, this is a breaking change to the beta SDK. I am working on releasing v10, which will include a new major version for all packages. Can I suggest you change the base branch for this PR from `master` to `release-v10`?
> 
> I can do that but as beta extends stable. when this is merged new beta users should be able to use sdk without any breaking change. `beta-sdk `has a `stable-sdk` dependency . am i wrong to think like that ? if so i can rebase it to that branch

@BugGambit can you share you opinion on this? also we know we are the only user andother consumers are waiting for beta to be moved to stable ?

**Comment by BugGambit** [comment]
@G0nza1es Ok, if you are sure you are the only users, then I happy with breaking it.
There are still some unresolved comments in this PR

File: packages/stable/src/__tests__/api/datamodels.int.spec.ts (hunk 7 above)
Comment (G0nza1es) [reply packages/stable/src/__tests__/api/datamodels.int.spec.ts]: I will revert it back. i think PR checks were failing and revert and see if it works

File: packages/stable/src/__tests__/testUtils.ts (hunk 10 above)
Comment (G0nza1es) [reply packages/stable/src/__tests__/testUtils.ts:225]: added spaces. should be with spaces

File: packages/stable/src/api/files/filesMultipartUploadSessionApi.ts (hunk 12 above)
Comment (G0nza1es) [reply packages/stable/src/api/files/filesMultipartUploadSessionApi.ts:28]: done it should not be more than requested limit 

`this.multiPartFileUploadResponse.uploadUrls.length - 1`

**Pull request merged**